src/components/Tables/Table.tsx

```diff
@@ -41,86 +40,59 @@ export interface ColumnStyleOptions {
+export interface SortOptions<Data, N extends keyof Data> {
+    key: N;
+    cmp: (a: Data[N], b: Data[N]) => number;
+}
+
+interface Props<Data> {
+    data: Data[];
+    headerKeysAndLabels: TableHeaders<Data>;
     checkboxes?: boolean;
     onRowSelection?: (rowIds: number[]) => void;
     reorderable?: boolean;
-    headerFilters?: string[];
+    filters?: (Filter<Data, string> | keyof Data)[];
     pagination?: boolean;
-    defaultShownHeaders?: string[];
-    toggleableHeaders?: string[];
-    sortable?: boolean;
+    defaultShownHeaders?: readonly (keyof Data)[];
+    toggleableHeaders?: readonly (keyof Data)[];
+    sortable?: (keyof Data)[];
     onEdit?: (data: number) => void;
     onDelete?: (data: number) => void;
-    columnDisplayFunctions?: TableColumnDisplayFunctions;
-    columnStyleOptions?: TableColumnStyleOptions;
-    onRowClick?: OnRowClickFunction;
+    columnDisplayFunctions?: { [headerKey in keyof Data]?: ColumnDisplayFunction<Data[headerKey]> };
+    columnStyleOptions?: { [headerKey in keyof Data]?: ColumnStyleOptions };
+    onRowClick?: OnRowClickFunction<Data>;
+    autoFilter?: boolean;
 }
 
-const doesRowIncludeFilterText = (
-    row: Row,
-    filterText: FilterText,
-    headers: TableHeaders
-): boolean => {
-    for (const [headerKey, _headerLabel] of headers) {
-        if (
-            !(row.data[headerKey] ?? "")
-                .toString()
-                .toLowerCase()
-                .includes((filterText[headerKey] ?? "").toLowerCase())
-        ) {
-            return false;
-        }
-    }
-    return true;
-};
-
-const dataToFilteredRows = (
-    data: Datum[],
-    filterText: FilterText,
-    headers: TableHeaders
-): Row[] => {
-    const rows = dataToRows(data, headers);
-    return filterRows(rows, filterText, headers);
-};
-
-const dataToRows = (data: Datum[], headers: TableHeaders): Row[] => {
-    return data.map((datum: Datum, currentIndex: number) => {
-        const row: Row = { rowId: currentIndex, data: { ...datum } };
-
-        for (const [headerKey, _headerLabel] of headers) {
-            const databaseValue = datum[headerKey] ?? "";
-            row.data[headerKey] = Array.isArray(databaseValue)
-                ? databaseValue.join(", ")
-                : databaseValue;
-        }
-
-        return row;
-    });
-};
-
-const filterRows = (rows: Row[], filterText: FilterText, headers: TableHeaders): Row[] => {
-    return rows.filter((row) => doesRowIncludeFilterText(row, filterText, headers));
-};
-
-interface CellProps {
-    row: Row;
-    columnDisplayFunctions: TableColumnDisplayFunctions;
-    headerKey: string;
+interface CellProps<Data> {
+    row: Row<Data>;
+    columnDisplayFunctions: { [headerKey in keyof Data]?: ColumnDisplayFunction<Data[headerKey]> };
+    headerKey: keyof Data;
 }
 
-const CustomCell: React.FC<CellProps> = ({ row, columnDisplayFunctions, headerKey }) => {
-    return (
+const CustomCell = <Data extends unknown>({
```

Comment:
can you explain why Data extends unknown?

Reply:
unknown is a fundamental type in typescript, and the type hierarchy goes `never <- unknown <- everythingElse`. It's essentially the same as any except with any you can perform operations on it like it's any other type.

In type parameters for arrow functions, you have to put some sort of type bound and saying that `Data` extends unknown means that there are no type constraints on `Data` aside from the fact that it can't be `never`